napari/utils/context/_service.py

```diff
@@ -0,0 +1,328 @@
+"""
+Classes that implement scoped contexts: dict-like objects that manage a set
+of key value pairs, inheriting from "parent" contexts when a value is missing.
+
+In most cases, `create_context` and `get_context` will be used to access the
+functionality here:
+
+Examples
+--------
+
+    from napari.utils.context._service import get_context, create_context
+
+    class A:
+        def __init__(self) -> None:
+            self.ctx = create_context(self)
+            self.b = B()  # self.b.ctx will be scoped off of self.ctx
+
+    class B:
+        def __init__(self) -> None:
+            self.ctx = create_context(self)
+
+    obj = A()
+    assert get_context(obj) is obj.ctx
+
+This module is modeled after vscode's context key service
+
+https://github.com/microsoft/vscode/blob/main/src/vs/platform/contextkey/browser/contextKeyService.ts
+"""
+from __future__ import annotations
+
+import sys
+import weakref
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from itertools import count
+from typing import Any, Callable, Dict, Optional, TypeVar
+
+from typing_extensions import Final
+
+from napari.utils.events import EventEmitter
+
+
+class Context(dict):
+    """A (hashable) dict that has a parent.
+
+    If `__getitem__` or `get` can't find a key, the parent is checked.
+    use `collect()` to get the merged dict (self takes precedence)
+    """
```

Comment:
I'm commenting as I'm reading so it might be answered later. 

Why not use ChainMap ? https://docs.python.org/3/library/collections.html#collections.ChainMap

Reply:
oh hell yes... thank you for reminding me.